### PR TITLE
[IMP] web: redesign notifications

### PR DESCRIPTION
**Pull request by anso-odoo**
This commit gives a fresh look to notification messages and makes them more consistent with new MILK design

task-3469789
enterprise PR (for test adaptation only): https://github.com/odoo/enterprise/pull/53590

_To test the notifications, copy paste in console:_
`function test() {console.log('test')};odoo.__WOWL_DEBUG__.root.env.services.notification.add('Hello world', {title:"Notification title", type: "info", buttons: [ {name: "Hey", primary: true, onClick: test}, {name: "Bye bye", primary: false, onClick: test} ] });`

| | Before | After |
|--------|--------|--------|
| Light mode | Screenshot 2023-12-18 at 15 24 35 | Screenshot 2023-12-18 at 15 23 31 |
| Dark mode | Screenshot 2023-12-18 at 15 24 44 | Screenshot 2023-12-18 at 15 23 49 |

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by anso-odoo** [comment]
Hello @stefanorigano ! Can you have a look at this and tell me what you think ? Thanks !

File: addons/web/static/src/core/notifications/notification.xml [hunk 1]
```diff
@@ -2,17 +2,15 @@
 <templates xml:space="preserve">
 
     <t t-name="web.NotificationWowl">
-        <div t-on-mouseenter="this.props.freeze" t-on-mouseleave="this.props.refresh" t-attf-class="o_notification {{props.className}} border border-{{props.type}} bg-view mb-2 position-relative"
+        <div t-on-mouseenter="this.props.freeze" t-on-mouseleave="this.props.refresh" t-attf-class="o_notification {{props.className}} bg-view border border-{{props.type}} mb-2 p-3 position-relative rounded shadow-lg"
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:5]: convention: background utility classes after border ones, and rounded immediately after border.
note that we should theoretically adapt the preexisting code by moving `position-relative` at the beginning... but let's not do that in order to keep the diff slim.

```suggestion
        <div t-on-mouseenter="this.props.freeze" t-on-mouseleave="this.props.refresh" t-attf-class="o_notification {{props.className}} border border-{{props.type}} rounded bg-view mb-2 p-3 position-relative shadow-lg"
```

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Hello and thanks @anso-odoo 👍 

About the visual result, I'm quite annoyed by the "border glitch" 😅
The issue was already there, but it's even more noticeable since you introduced the gray border.
image.


I have two proposals, I think that the "new element" is the best, wdyt?

| . | . | . |
|--------|--------|--------|
| This PR | image | . |
| New element | image | image |
| Box Shadow | image | image | 

Some comments for you in the code... cheers!

File: addons/web/static/src/core/notifications/notification.xml [hunk 2]
```diff
@@ -2,17 +2,15 @@
...
-            <div t-attf-class="o_notification_body ps-3 pe-5 py-2">
+            <h5 t-if="props.title" class="o_notification_title d-block m-0 pe-5" t-out="props.title"/>
+            <button type="button" class="o_notification_close btn-close" aria-label="Close" t-on-click="props.close"/>
+            <div t-attf-class="o_notification_body {{props.title ? 'mt-2' : ''}} pe-5">
                 <div t-if="props.message" class="me-auto o_notification_content" t-out="props.message"/>
-                <div t-if="props.buttons.length" class="o_notification_buttons d-flex gap-2 mt-2">
+                <div t-if="props.buttons.length" class="o_notification_buttons d-flex gap-2 mt-3">
                     <button t-foreach="props.buttons" t-as="button" type="button" t-key="button_index"
-                    t-attf-class="btn {{button.primary ? 'btn-primary' : 'btn-secondary'}}" t-on-click="button.onClick">
+                    t-attf-class="btn {{button.primary ? 'btn-' + props.type : 'btn-secondary'}}" t-on-click="button.onClick">
```
[the remark this answers is not in the document]
Comment (anso-odoo) [reply addons/web/static/src/core/notifications/notification.xml:13]: Imo the "Cancel/Abort" should always be the secondary option: 

If the user wants to perform what we see as a really critical action (e.g.: destroying something) we have to assume it's something he is trying to do and that he has a reason to do it, because the application is allowing him to do it. So, the warning "Are you sure you want to destroy" -> main action should be "Yes, destroy"

[the remark this answers is not in the document]
Comment (stefanorigano) [reply addons/web/static/src/core/notifications/notification.xml:13]: I'm sorry, the example provided is debatable at least.

Historically we "guide" the user to the safest option by using primary action... but it doesn't fit every scenario and having a danger `["Yes, destroy"]` button would make much more sense indeed, you're right 👍

I think that I was misled by the change in paradigm you're introducing, basically saying that "`primary` ≠ _primary color_" anymore, but "`primary` = _action we expect the user to perform_". 
This changes of perspective highlight a little paradox: 

---

**You destroyed everything**.       

[Undo]

---

[Undo] cannot be red (primary) of course... we could argue that it should be "secondary", but "_secondary to what?_" since these words don't means colors anymore but actions' hierarchy ? 

To summarize, having danger/warning... buttons is ok to me, but the wording and the implementation is misleading.
Does it make sense?

**Comment by anso-odoo** [comment]
Hello @stefanorigano. Tests are finally green. Can you have a look please ?

File: addons/web/static/src/core/notifications/notification.xml [hunk 3]
```diff
@@ -2,25 +2,26 @@
...
-                    t-attf-class="btn {{button.primary ? 'btn-primary' : 'btn-secondary'}}" t-on-click="button.onClick">
-                    <t t-if="button.icon">
-                        <i t-if="button.icon.indexOf('fa-') === 0" role="img"
-                        t-att-aria-label="button.name" t-att-title="button.name"
-                        t-attf-class="fa fa-fw o_button_icon {{button.icon}}"/>
-                        <img t-else="" t-att-src="button.icon" t-att-alt="button.name"/>
-                    </t>
-                    <span t-esc="button.name"/>
-                </button>
+            <div t-attf-class="o_notification_border bg-{{props.type}} rounded-start"/>
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:7]: Could you use a `<span>` tag instead?
Semantically it would be a better match for such a generic decorative element.

File: addons/bus/static/tests/simple_notification_tests.js [hunk 4]
```diff
@@ -44,7 +44,7 @@ QUnit.test("receive and display simple notification with specific type", async (
         message: "simple notification",
         type: "info",
     });
-    await contains(".o_notification.border-info");
+    await contains(".o_notification:has(.o_notification_border.bg-info)");
```
Comment (stefanorigano) [inline addons/bus/static/tests/simple_notification_tests.js:47]: Isn't
```js
await contains(".o_notification_border.bg-info");
```
enough? 🤔

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Thanks @anso-odoo 👍 
Some notes in the code.

Since you increased the padding (and it looks great on desktop), we now have an issue on mobile tho.
The padding is not taken in account by the `$o-notification-max-width` variable, especially 'cause it's used to compute the position rather than explicitly setting its width.

Could you review that? Thanks! 

image

File: addons/web/static/src/core/notifications/notification.xml [hunk 5]
```diff
@@ -2,25 +2,26 @@
...
+                        t-attf-class="btn {{button.primary ? 'btn-primary' : 'btn-secondary'}}" t-on-click="button.onClick">
+                        <t t-if="button.icon">
+                            <i t-if="button.icon.indexOf('fa-') === 0" role="img"
+                            t-att-aria-label="button.name" t-att-title="button.name"
+                            t-attf-class="fa fa-fw o_button_icon {{button.icon}}"/>
+                            <img t-else="" t-att-src="button.icon" t-att-alt="button.name"/>
+                        </t>
+                        <span t-esc="button.name"/>
+                    </button>
+                    </div>
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:24]: Indentation issue 😉

File: addons/web/static/src/core/notifications/notification.xml [hunk 6]
```diff
@@ -2,25 +2,26 @@
...
-                </button>
+            <span t-attf-class="o_notification_border bg-{{props.type}} rounded-start"/>
+            <div class="w-100 py-3 ps-3 pe-5 border border-start-0 rounded-end">
+                <h5 t-if="props.title" class="o_notification_title m-0" t-out="props.title"/>
+                <button type="button" class="o_notification_close btn-close position-absolute top-0 end-0 mt-3 me-3" aria-label="Close" t-on-click="props.close"/>
+                <div class="o_notification_body pe-5" t-att-class="{'mt-2' : props.title}">
+                    <div t-if="props.message" class="me-auto o_notification_content" t-out="props.message"/>
+                    <div t-if="props.buttons.length" class="o_notification_buttons d-flex gap-2 mt-3">
+                        <button t-foreach="props.buttons" t-as="button" type="button" t-key="button_index"
+                        t-attf-class="btn {{button.primary ? 'btn-primary' : 'btn-secondary'}}" t-on-click="button.onClick">
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:15]: I know that the file was already indentated weirdly, but since we're modifying these lines we should fix it 😉

File: addons/web/static/src/core/notifications/notification.xml [hunk 7]
```diff
@@ -2,25 +2,26 @@
...
+                        <t t-if="button.icon">
+                            <i t-if="button.icon.indexOf('fa-') === 0" role="img"
+                            t-att-aria-label="button.name" t-att-title="button.name"
+                            t-attf-class="fa fa-fw o_button_icon {{button.icon}}"/>
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:19]: `o_button_icon` is deprecated, we can remove it 😉

**Review by stefanorigano — CHANGES_REQUESTED** [review]
Very good @anso-odoo , thanks 🙌
Visual result it's much better, I've just some minor remarks about the code.
We're almost there, courage 💪

File: addons/web/static/src/core/notifications/notification.xml [hunk 8]
```diff
@@ -2,25 +2,22 @@
...
+            <span t-attf-class="o_notification_border bg-{{props.type}} rounded-start"/>
+            <div class="w-100 py-3 ps-3 pe-5 border border-start-0 rounded-end">
+                <h5 t-if="props.title" class="o_notification_title m-0" t-out="props.title"/>
+                <button type="button" class="o_notification_close btn-close position-absolute top-0 end-0 mt-3 me-3" aria-label="Close" t-on-click="props.close"/>
+                <div class="o_notification_body pe-5" t-att-class="{'mt-2' : props.title}">
+                    <div t-if="props.message" class="me-auto o_notification_content" t-out="props.message"/>
+                    <div t-if="props.buttons.length" class="o_notification_buttons d-flex gap-2 mt-3">
+                        <button t-foreach="props.buttons" t-as="button" type="button" t-key="button_index" t-attf-class="btn {{button.primary ? 'btn-primary' : 'btn-secondary'}}" t-on-click="button.onClick">
+                            <t t-if="button.icon">
+                                <i t-if="button.icon.indexOf('fa-') === 0" role="img" t-att-aria-label="button.name" t-att-title="button.name" t-attf-class="fa fa-fw o_button_icon"/>
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.xml:15]: `{{button.icon}}` is missing thus no icon is rendered ;)
In a previous comment I said that `o_button_icon` class is deprecated and could be removed, maybe you confused the two?

File: addons/web/static/src/core/notifications/notification.scss [hunk 9]
```diff
@@ -3,16 +3,16 @@
...
-        border-left-width: $o-notification-border-left-width !important; // Overrides border-[x] utility classes
-        box-shadow: $o-notification-shadow;
+    @include media-breakpoint-up(sm) {
+        left: auto;
+        width: $o-notification-max-width-md;
     }
 
-    .o_notification_close {
-        @include o-position-absolute(0, 0);
+    .o_notification_border {
```
Comment (stefanorigano) [inline addons/web/static/src/core/notifications/notification.scss:14]: I'm wondering if we should use another name for this element and its variable... each time I check this PR I'm mislead to think that's about the actual card border.

 `o_notification_flag`?  `o_notification_bar`? What do you think?

**Review by stefanorigano — CHANGES_REQUESTED** [review]

File: addons/web/static/src/core/notifications/notification.xml (hunk 8 above)
Comment (anso-odoo) [reply addons/web/static/src/core/notifications/notification.xml:15]: Indeed I mingled both 
Added a margin class as a counterpart

File: addons/web/static/src/core/notifications/notification.scss (hunk 9 above)
Comment (anso-odoo) [reply addons/web/static/src/core/notifications/notification.scss:14]: notification bar seems better indeed 👍 
Thanks !

File: addons/product/static/tests/product_pricelist_report_test.js [hunk 10]
```diff
@@ -121,10 +122,11 @@ QUnit.module('Product Pricelist Report', {
         await editInput(inputElement, '', 4);
         await click(addQtyButton);
         notificationElement = document.body.querySelector('.o_notification_manager .o_notification:nth-child(2)');
+        notificationElementBar = document.body.querySelector('.o_notification_manager .o_notification:nth-child(2) .o_notification_bar');
         assert.strictEqual(notificationElement.querySelector('.o_notification_content').textContent,
             "At most 5 quantities can be displayed simultaneously. Remove a selected quantity to add others.",
             "Can not add more then 5 quantities");
-        assert.hasClass(notificationElement, "border-warning");
+            assert.hasClass(notificationElementBar, "bg-warning");
```
Comment (Feyensv) [inline addons/product/static/tests/product_pricelist_report_test.js:129]:
```suggestion
        assert.hasClass(notificationElementBar, "bg-warning");
```

**Comment by anso-odoo** [comment]
Hello @aab-odoo this is ready for tech review, can you have a look ? Thanks !
(I'm assuming you can review this as you reviewed the [past redesign](https://github.com/odoo/odoo/pull/82524#issuecomment-1015259654))

**Review by Julien00859 — APPROVED** [review]
review for bus

**Comment by anso-odoo** [comment]
Hello @aab-odoo, indeed the screenshots are now outdated. I'll update that part to avoid confusion. Thanks !

File: addons/web/static/src/core/notifications/notification.scss [hunk 11]
```diff
@@ -3,16 +3,16 @@
...
-    @include o-position-absolute($o-navbar-height * 1.15, map-get($spacers, 2), $left: calc(100vw - #{$o-notification-max-width}));
+    @include o-position-absolute($top: $o-navbar-height * 1.15, $right: $o-notification-margin, $left: $o-notification-margin);
     z-index: $o-notification-zindex;
 
-    .o_notification {
-        border-left-width: $o-notification-border-left-width !important; // Overrides border-[x] utility classes
-        box-shadow: $o-notification-shadow;
+    @include media-breakpoint-up(sm) {
+        left: auto;
+        width: $o-notification-max-width-md;
```
Comment (pparidans) [inline addons/web/static/src/core/notifications/notification.scss:11]: :warning: you use both `sm` and `md` breakpoint name together !

File: addons/bus/static/tests/simple_notification_tests.js [hunk 12]
```diff
@@ -44,7 +44,7 @@ QUnit.test("receive and display simple notification with specific type", async (
         message: "simple notification",
         type: "info",
     });
-    await contains(".o_notification.border-info");
+    await contains(".o_notification:has(.o_notification_bar.bg-info)");
```
Comment (pparidans) [inline addons/bus/static/tests/simple_notification_tests.js:47]: :warning: `:has` is still not supported by a stable version of Firefox (120 as of this review).
It would then make those tests break in that browser.

**Update**: you're very lucky: Firefox 121 (which supports `:has`) has just been released today! :nerd_face:

**Review by pparidans — CHANGES_REQUESTED** [review]
The community build is :green_circle: but not the **enterprise** one :red_circle: !

**Review by pparidans — APPROVED** [review]
LGTM :+1: 

@anso-odoo Could you create the enterprise PR, please ?

**Comment by anso-odoo** [comment]
Hello @pparidans. Here is the enterprise branch https://github.com/odoo/enterprise/pull/53590. Runbot is red but the two errors are "already known"... Should I do something about that ? 

Thanks a lot for checking it 👍

**Review by pparidans — APPROVED** [review]
@robodoo r+
